internal/state/apply.go

```diff
@@ -2,44 +2,35 @@ package state
 
 import (
 	"context"
-
 	"github.com/kyma-project/serverless-manager/api/v1alpha1"
 	"github.com/kyma-project/serverless-manager/internal/chart"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 // run serverless chart installation
-func sFnApplyResources() stateFn {
-	return func(ctx context.Context, r *reconciler, s *systemState) (stateFn, *ctrl.Result, error) {
-		// check if condition exists
-		if !s.instance.IsCondition(v1alpha1.ConditionTypeInstalled) {
-			return nextState(
-				sFnUpdateProcessingState(
-					v1alpha1.ConditionTypeInstalled,
-					v1alpha1.ConditionReasonInstallation,
-					"Installing for configuration",
-				),
-			)
-		}
-
-		// install component
-		err := chart.Install(s.chartConfig)
-		if err != nil {
-			r.log.Warnf("error while installing resource %s: %s",
-				client.ObjectKeyFromObject(&s.instance), err.Error())
-			return nextState(
-				sFnUpdateErrorState(
-					v1alpha1.ConditionTypeInstalled,
-					v1alpha1.ConditionReasonInstallationErr,
-					err,
-				),
-			)
-		}
+func sFnApplyResources(ctx context.Context, r *reconciler, s *systemState) (stateFn, *ctrl.Result, error) {
+	// check if condition exists
+	if !s.instance.IsCondition(v1alpha1.ConditionTypeInstalled) {
+		s.setState(v1alpha1.StateProcessing)
+		s.instance.UpdateConditionUnknown(v1alpha1.ConditionTypeInstalled, v1alpha1.ConditionReasonInstallation,
+			"Installing for configuration")
+		return nextState(sFnUpdateStatusAndRequeue)
+	}
 
-		// switch state verify
-		return nextState(
-			sFnVerifyResources(),
+	// install component
+	err := chart.Install(s.chartConfig)
+	if err != nil {
+		r.log.Warnf("error while installing resource %s: %s",
+			client.ObjectKeyFromObject(&s.instance), err.Error())
+		setErrorState(s,
+			v1alpha1.ConditionTypeInstalled,
+			v1alpha1.ConditionReasonInstallationErr,
+			err,
 		)
```

Comment:
I'm a little confused why in some cases we use the `setErrorState` method and in other cases, we use methods to update state and condition separately, like here:

```
s.setState(v1alpha1.StateProcessing)
s.instance.UpdateConditionUnknown(v1alpha1.ConditionTypeInstalled, v1alpha1.ConditionReasonInstallation,
			"Installing for configuration")
```

Does it make sense to unify these parts? or maybe It makes sense to keep it as it is? It's an open question I can't decide which solution is better

Reply:
In my opinion methods like the `setErrorState` consume less code and help with code aggregation, but having separated methods helps with keeping code clear. The only problem I have with the set error is that the name suggests that it would update state only, but what about conditions?